### DISCUSSION TITLE
位元運算子的應用

### DISCUSSION THREAD

**Discussion by ThomasMeng0515** [1101-GS4719-C/Python-Community-GS4719-C#452] [General]
# 位元運算子
今天上課的時候老師有講解到了位元運算子，老師也提及了一些位元運算子的應用，而大部分的應用皆屬於硬體的範疇，在這裡也想跟大家分享另一個位元運算子的硬體上之應用-子網路遮罩(subnet mask)。

## 子網路遮罩與 & 運算子
子網路遮罩的功能簡單來說就是決定哪些 IP 位置是屬於同一個子網路用的。
常見的一個子網路遮罩為: **255.255.255.0** 
用下面三個IP位址當作例子:

- **192.168.1.1**

- **192.168.1.123**

- **192.168.2.1**

我們將子網路遮罩以及 IP位址皆轉換為二進制(binary)表示法如下:
![20128159hsiJP4n8C2](https://user-images.githubusercontent.com/99859555/157455832-36c63bef-7397-4e5a-9226-fd75fac38dbd.jpg)

再將三個 IP位址各別對子網路遮罩作 & 運算，可以得到:
![20128159X3P3UCwoVb](https://user-images.githubusercontent.com/99859555/157456024-e09a952e-6a6f-4c01-a4f1-eba56d6a029d.jpg)

從結果來看，第一個與第二個 IP位址都是同一個結果：**192.168.1.0**，第三個 IP位址則得到：**192.168.2.0**。
因此我們可以知道，**前兩個 IP位址屬於同一個子網路，而第三個 IP位址則是屬於其它子網路**。
從這個操作上來看把 IP位址與子網路作 & 運算就好像丟到一個遮罩裡面把不要的資訊蓋掉，所以遮罩的名字便從此而來。

## 參考資料
[理解網際網路協定（三）：何謂「子網路遮罩」，什麼又是「預設閘道」？](https://ithelp.ithome.com.tw/articles/10272208/)

**Comment by awkward-willy** [comment]
這裡順便補充一些常見應用：
1. 奇偶數判斷
常見的做法是%2，若結果為1則為奇數，反之則為偶數
但也可以 &1，若結果為1則為奇數，反之則為偶數
因為奇數的二進位形式必定為......1 經 AND 運算後必為1

2. 多次乘 / 除以2
左移與右移可以讓原數分別 *2 或 /2（取較小值）

3. 紀錄是非值
若欲紀錄多筆資料，每一筆皆只分成是與否（有或無......），則可以用二進位的方式進行紀錄，節省記憶體空間
如：紀錄1名學生8學期內是否有申請宿舍，則可以用 11111111(255) 記錄，相比於用 List = [1, 1, 1, 1, 1, 1, 1, 1]（或其他語言的陣列）小上許多

4. 完成兩變數的 Swap
可以利用 XOR 的性質（XOR 兩次後會得原本值）完成變數的交換
如：a = 5, b= 10；經 a = a^b 和 b = b^a 與 a = a^b 可得 a = 10, b = 5

應該還有許多用途，小弟我暫時想不到了，也歡迎各大大補充！

**Reply by ThomasMeng0515** [reply]
在硬體設計上利用左移(<<)、右移(>>)計算2的冪次方乘除法也會比較節省硬體資源，較有效率。

**Comment by sean1325418** [comment]
硬體邏輯都是用位元1、0來表示，想搞懂硬體的話最好先從位元運算學起

**Comment by Yuhsin000** [comment]
謝謝，多學了很多個運算子與位元的應用；另外也謝謝awkward-willy，我覺得第三點很實用

**Comment by Ku-Hao-Chen** [comment]
謝謝分享。

**Comment by xkllkx** [comment]
順帶一提，python也能做進制轉換喔！[python進制轉換](https://clay-atlas.com/blog/2020/09/24/python-cn-convert-bin-oct-dec-hex/)

**Comment by boson13579** [comment]
這邊再補充一個位元運算實用的東東好了
比如說我想算 n^m % p
如果直接算得話就要 至少 O(m) 的雜度
不過我們可以把code改成這樣
def QuickPower(n , m, p):
    ans = 1
    while(m != 0):
        if m&1:
            ans = (ans*n)%p
        n = n*n%p
        m >>= 1
    return ans
這樣time complexity 就只要O(log(m))喔,快很多ㄅ

**Comment by RogelioKG** [comment]
樓上 @boson13579 提到的是一種叫快速冪的技巧，
能夠用對數級的時間複雜度實現求冪，
這邊卡一個之前寫的 C++ 版。
這裡面就有用到位元運算子。

```c++
long long fast_pow(long long base, long long power)
{
    long long val = 1;
    while (power)
    {
        if (power & 1)
            val *= base;
        power >>= 1;
        base *= base;
    }
    return val;
}
```

**Comment by RogelioKG** [comment]
當然矩陣的相乘也可以使用快速冪。
再卡一個之前寫的，矩陣快速冪：
```c++
struct Mat
{
    long long mat[2][2];

    Mat operator*(Mat & m) const
    {
        Mat ret;
        for (int i = 0; i < 2; i++)
            for (int j = 0; j < 2; j++)
                for (int k = 0; k < 2; k++)
                    ret.mat[i][j] += mat[i][k] * m.mat[k][j];
        return ret;
    }
};

Mat fast_pow(Mat base, long long power)
{
    Mat result {1,0,0,1}; // 單位矩陣
    while (power)
    {
        if (power & 1)
            result = result * base;
        power >>= 1;
        base = base * base;
    }
    return result;
}
```
